pymmcore_plus/mda/_engine.py

```diff
@@ -0,0 +1,113 @@
+import time
+from typing import TYPE_CHECKING, Protocol, runtime_checkable
+
+from loguru import logger
+from useq import MDASequence
+
+from .events import PMDASignaler, _get_auto_MDA_callback_class
+
+if TYPE_CHECKING:
+    from ..core import CMMCorePlus
+
+
+@runtime_checkable
+class PMDAEngine(Protocol):
```

Comment:
`Protocol` also actually works as an ABC.  So, if we add some `@abstractmethod` decorators here, someone could (optionally) use it as a base class and make sure their protocol is in spec.  It still supports structural sub-typing though, so they don't have to use.

So this could look like 

```python

@runtime_checkable
class PMDAEngine(Protocol):
    @abstractmethod
    def __init__(self, mmc: "CMMCorePlus") -> None:
        ...

    @abstractproperty
    def events(self) -> PMDASignaler:
        """Add docstring prescribing behavior."""

    @abstractmethod
    def cancel(self):
        """Add docstring prescribing behavior."""

    @abstractmethod
    def toggle_pause(self):
        """Add docstring prescribing behavior."""

    @abstractmethod
    def run(self, sequence: MDASequence):
        """Add docstring prescribing behavior."""
```

and then lower (optional, but nice to immediately show that it implements the protocol)

```python
class MDAEngine(PMDAEngine):
```

Reply:
oh i re-read what you wrote. I skimmed too quickly

Reply:
> I think it might also be nice to include a method like the waiting loop as a method that can be used by a subclass. Is it ok to have a method with implementation on the ABC or should that really go somewhere else?

yeah, it's technically ok... but if you're imagining something like a hook that gets called somewhere inside of `run`, then i think it actually makes the "protocol" part a lot more complicated.  Since whether that hook gets called or not really depends on whether the `run` method calls it: making it more of an implementation detail of the engine, rather than an enforceable part of the protocol